### PR TITLE
Merge safe features branch

### PR DESCRIPTION
**Pull request by EmelyanenkoK**

### CODE REVIEW

**Review by Elgayar777 — COMMENTED** [review]
0k

**Comment by GrozneyS** [comment]
> 0k

Ok

**Review by GrozneyS — COMMENTED** [review]
Ok

**Review by Alihavaej — COMMENTED** [review]
Kid

**Comment by Volimjasnu12** [comment]
Ok

**Review by yyy7u — APPROVED** [review]

**Review by yyy7u — APPROVED** [review]

**Review by Meroksa — APPROVED** [review]

**Review by Meroksa — APPROVED** [review]

File: adnl/adnl-channel.cpp [hunk 1]
```diff
@@ -112,16 +112,16 @@ void AdnlChannelImpl::send_message(td::uint32 priority, td::actor::ActorId<AdnlN
...
-          td::actor::send_closure(peer, &AdnlPeerPair::receive_packet_from_channel, channel_id, std::move(packet));
-        }
-      });
+  auto P = td::PromiseCreator::lambda([peer = peer_pair_, channel_id = channel_in_id_, addr, id = print_id(),
+                                       size = data.size()](td::Result<AdnlPacket> R) {
+    if (R.is_error()) {
+      VLOG(ADNL_WARNING) << id << ": dropping IN message: can not decrypt: " << R.move_as_error();
+    } else {
+      auto packet = R.move_as_ok();
+      packet.set_remote_addr(addr);
```
Comment (Meysamlak) [inline adnl/adnl-channel.cpp:121]: UQDLMomHM8nqeqraWC4wm-T_6gXsDTG_ER1XtW9TB6ZV0KR1

File: adnl/adnl-local-id.cpp [hunk 2]
```diff
@@ -41,20 +41,34 @@ AdnlAddressList AdnlLocalId::get_addr_list() const {
...
-          td::actor::send_closure(peer_table, &AdnlPeerTable::receive_decrypted_packet, dst, std::move(packet));
-        }
-      });
-
+  InboundRateLimiter& rate_limiter = inbound_rate_limiter_[addr];
+  if (!rate_limiter.rate_limiter.take()) {
+    VLOG(ADNL_NOTICE) << this << ": dropping IN message: rate limit exceeded";
+    add_dropped_packet_stats(addr);
+    return;
+  }
```
Comment (Meysamlak) [inline adnl/adnl-local-id.cpp:49]: UQDLMomHM8nqeqraWC4wm-T_6gXsDTG_ER1XtW9TB6ZV0KR1

**Review by Meysamlak — COMMENTED** [review]
UQDLMomHM8nqeqraWC4wm-T_6gXsDTG_ER1XtW9TB6ZV0KR1

File: adnl/adnl-local-id.cpp (hunk 2 above)
Comment (Meysamlak) [reply adnl/adnl-local-id.cpp:49]: Please, if you can deposit my money, Bakhtar will not enter my country

**Comment by Hamidasl** [comment]
1bef6df45513075d5485dd5b6b69174cd6855401

**Review by Fart10 — COMMENTED** [review]
👹

**Review by robinkoms — COMMENTED** [review]
Ok

File: adnl/adnl-channel.cpp (hunk 1 above)
Comment (flores86ryan) [reply adnl/adnl-channel.cpp:121]: Smile

**Review by Hyperdello — APPROVED** [review]

**Comment by ayanwaar** [comment]
> 0k

[]()

**Review by Hyperdello — APPROVED** [review]

File: adnl/adnl-channel.cpp (hunk 1 above)
Comment (Jones098) [reply adnl/adnl-channel.cpp:121]: Yes

**Review by MrG9090 — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]

**Review by burkedaniel205 — COMMENTED** [review]
Beautiful

**Review by Hyperdello — APPROVED** [review]

**Review by MrG9090 — APPROVED** [review]

**Review by MrG9090 — APPROVED** [review]

**Comment by burkedaniel205** [comment]
Private key needs to be used

**Review by Hyperdello — APPROVED** [review]

**Review by MrG9090 — APPROVED** [review]

**Comment by MOHAMAD1313M** [comment]
Hello, don't be shy. I wanted to know if you have any educational videos for this program so we can better understand and implement it.

**Comment by MrG9090** [comment]
257a716a17f2aa7ea33ec440fe0d918f88b3fbaf

**Review by MrG9090 — APPROVED** [review]

**Review by burkedaniel205 — COMMENTED** [review]
K

**Pull request merged**